Makefile

```diff
@@ -23,7 +23,19 @@ install:
 	${OPAM} reinstall cooltt
 
 test:
-	@${DUNE} build @install @runtest
+	@${DUNE} build --display=quiet @runtest @test-display
+
+# Here we want full reproducible behavior, ideally, we build all the
+# dependencies of the tests before actually running them, as to avoid
+# interference. This is possible to do cleanly in Dune 3.0 as `alias`
+# got the right semantics, so you can collect all of your tests deps
+# in an alias and use `dune build @test-deps` instead of the
+# hand-setup below.
+TEST_FILE=test/test.exe
+test-timings:
+	@${DUNE} clean --display=quiet
+	@${DUNE} build --display=quiet $(TEST_FILE) $(shell ls test/*.cooltt test/cooltt-lib)
```

Comment:
One thing I (think I) would like is to be able to run the tests fresh _without_ running the build fresh. My use-case is to be able to run something like `time run the tests` and get the amount of time it takes to run the tests, not the amount of time it takes to build the whole project. Is that reasonable, or do you think that's a bad idea?

P.S. I really appreciate your help. I have to say that while I know how to use OCaml pretty well, I have found it very difficult to learn the build tools and I am really grateful for your help and patience.

Reply:
You are most welcome!

You could in that case just clean the output file, then dune will recreate it, but this is still fragile, unless you give up incrementality. By fragile I mean, it is hard to ensure you are in the right state in the build DAG.

As you got a single test runner you can just run it directly if you want, that is to say, outside of dune. Tho, if you are serious about timing still that will provide little help, as noise will be usually too high; YMMV, maybe you could be interested in some bench-specific packages such as core-bench which you could integrate inside the test runner .

Dune is a system that allows you to declare a DAG as the specification of incremental computation, where arrows a pure functions from nodes [files usually, tho other stuff such as env vars is also supported], and will just try to build the desired node as fast a possible, and in a sound way.

Incremental computation is indeed in conflict with the requirements of benchmarking , as for example it is common in a bench to do N equal runs and average, whereas the whole purpose of tools like dune is actually to try to cache that exact build operation.